coderd/workspacebuilds_test.go

```diff
@@ -1151,3 +1152,118 @@ func TestMigrateLegacyToRichParameters(t *testing.T) {
 	require.Len(t, buildParameters, 1)
 	require.Equal(t, "carrot", buildParameters[0].Value)
 }
+
+func TestWorkspaceBuildDebugMode(t *testing.T) {
+	t.Parallel()
+
+	t.Run("AsRegularUser", func(t *testing.T) {
+		t.Parallel()
+
+		// Create users
+		templateAuthorClient := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
+		templateAuthor := coderdtest.CreateFirstUser(t, templateAuthorClient)
+		regularUserClient, _ := coderdtest.CreateAnotherUser(t, templateAuthorClient, templateAuthor.OrganizationID)
+
+		// Template owner: create a template
+		version := coderdtest.CreateTemplateVersion(t, templateAuthorClient, templateAuthor.OrganizationID, nil)
+		template := coderdtest.CreateTemplate(t, templateAuthorClient, templateAuthor.OrganizationID, version.ID)
+		coderdtest.AwaitTemplateVersionJob(t, templateAuthorClient, version.ID)
+
+		// Regular user: create a workspace
+		workspace := coderdtest.CreateWorkspace(t, regularUserClient, templateAuthor.OrganizationID, template.ID)
+		coderdtest.AwaitWorkspaceBuildJob(t, regularUserClient, workspace.LatestBuild.ID)
+
+		// Regular user: try to start a workspace build in debug mode
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
+
+		_, err := regularUserClient.CreateWorkspaceBuild(ctx, workspace.ID, codersdk.CreateWorkspaceBuildRequest{
+			TemplateVersionID: workspace.LatestBuild.TemplateVersionID,
+			Transition:        codersdk.WorkspaceTransitionStart,
+			LogLevel:          "debug",
+		})
+
+		// Regular user: expect an error
+		require.NotNil(t, err)
+		var sdkError *codersdk.Error
+		isSdkError := xerrors.As(err, &sdkError)
+		require.True(t, isSdkError)
+		require.Contains(t, sdkError.Message, "Workspace builds with a custom log level are restricted to template authors only.")
+	})
+	t.Run("AsTemplateAuthor", func(t *testing.T) {
+		t.Parallel()
+
+		client := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
+		user := coderdtest.CreateFirstUser(t, client)
```

Comment:
Blocking: instead of running this as the "admin" user, we should instead do as above and create another user with template admin permissions. This would more properly test that the template author has the correct permissions to create this sort of job.

Reply:
Thank you for the hint! Refactored.